### PR TITLE
Issue 903: Delete other contacts (DK sandbox)

### LINKED ISSUE #903: Registrar: Add and implement button to "Remove extra contacts"
### Story

As a applicant
I want to be able to remove extra contacts I've added on the [Other employees from your org page](https://getgov-staging.app.cloud.gov/register/other_contacts/)
so that I don't waste time erasing the data I entered into the organization contact fields

### Acceptance Criteria
- [ ] fix add button to match the functionality on nameservers, dnssec and so on
- [ ] implement buttons to "remove" extra contacts on Other Contacts page, [using same pattern](https://github.com/cisagov/getgov/pull/1166) from DNSSEC and Name Servers.
- [ ] Implement look and functionality from this [Figma mockup](https://www.figma.com/file/aJbuDLJNNikqQObTuCNKQa/GetGov_Designs_ECS-Truss?type=design&mode=design). To enter prototype mode click the "present" button at the top right.

### Additional Context

See [Slack Thread](https://cisa-corp.slack.com/archives/C05BDEA3C11/p1697669491409489)

### Issue Links

🔄 Relates to #589

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #903 

## Changes

- Fixed add button's functionality
- Added a delete button that works similar to DNSSEC and Name Servers'
- Unit test new backend features (delete widget on formset)

### UX

Cleared with @gabydisarli 

Assuming you have an application with other contacts and you are hitting the 'delete' buttons, once you have one contact left on the page, hitting its 'delete' will auto-fill the "No other contacts" radio form and switch you over to the no_other_contacts_rationale textbox to fill in. Hitting "Yes" on the radio form switches you back to the view with the last contact.

Assuming you select "Yes" on a new application and hit submit without filling in a contact, you should get validation errors.

Assuming you are on an application that has a contact, you add an empty form in the formset and delete the existing contact, hit submit, you get the same validation errors as in the above scenario.

### Dev

This is radically different than the implementation on DNSSEC and Name Servers. In the latter 2, we just remove the deleted forms fro the page and rely on the model to do a diff and delete what's missing from Cloudflare.

In this case, we use Django's deletion_widget (hidden) on the formset and use JS to hide forms marked for deletion instead of removing them. We only worry about dynamically updating the labels (as opposed to the input names etc.) and to do so we look at the visible forms, not the total forms. 

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/assets/js/get-gov.js [hunk 1]
```diff
@@ -229,127 +229,245 @@ function handleValidationClick(e) {
...
-            // Replace the attribute value with the updated value
-            attr.value = attr.value.replace(formNumberRegex, `form-${index}-`);
-          }
-        });
+function removeForm(e, formLabel, isNameserversForm, addButton, formIdentifier){
+  let totalForms = document.querySelector(`#id_${formIdentifier}-TOTAL_FORMS`);
+  let formToRemove = e.target.closest(".repeatable-form");
+  formToRemove.remove();
+  let forms = document.querySelectorAll(".repeatable-form");
+  totalForms.setAttribute('value', `${forms.length}`);
```
Comment (zandercymatics) [inline src/registrar/assets/js/get-gov.js:241]: Nice refactor

File: src/registrar/assets/js/get-gov.js [hunk 2]
```diff
@@ -229,127 +229,245 @@ function handleValidationClick(e) {
...
   let formLabel = '';
   let isNameserversForm = document.title.includes("DNS name servers |");
+  let isOtherContactsForm = document.title.includes("Other employees from your organization");
+  // The Nameservers formset features 2 required and 11 optionals
   if (isNameserversForm) {
     cloneIndex = 2;
     formLabel = "Name server";
-  } else if ((document.title.includes("DS Data |")) || (document.title.includes("Key Data |"))) {
+  // DNSSEC: DS Data
+  } else if (document.title.includes("DS Data |")) {
```
Comment (zandercymatics) [inline src/registrar/assets/js/get-gov.js:453]: Same concern as before, though I think this should be moved into a variable

File: src/registrar/assets/js/get-gov.js [hunk 3]
```diff
@@ -360,7 +478,7 @@ function prepareDeleteButtons(formLabel) {
       let forms = document.querySelectorAll(".repeatable-form");
       let formNum = forms.length;
       let newForm = repeatableForm[cloneIndex].cloneNode(true);
-      let formNumberRegex = RegExp(`form-(\\d){1}-`,'g');
+      let formNumberRegex = RegExp(`${formIdentifier}-(\\d){1}-`,'g');
```
Comment (zandercymatics) [inline src/registrar/assets/js/get-gov.js:481]: Nice refactor

File: src/registrar/assets/js/get-gov.js [hunk 4]
```diff
@@ -393,16 +511,27 @@ function prepareDeleteButtons(formLabel) {
       }
 
       formNum++;
-      newForm.innerHTML = newForm.innerHTML.replace(formNumberRegex, `form-${formNum-1}-`);
-      newForm.innerHTML = newForm.innerHTML.replace(formLabelRegex, `${formLabel} ${formNum}`);
+
+      newForm.innerHTML = newForm.innerHTML.replace(formNumberRegex, `${formIdentifier}-${formNum-1}-`);
+      // For the other contacts form, we need to update the fieldset headers based on what's visible vs hidden,
+      // since the form on the backend employs Django's DELETE widget. For the other formsets, we delete the form
+      // in JS (completely remove from teh DOM) so we update the headers/labels based on total number of forms.
```
Comment (zandercymatics) [inline src/registrar/assets/js/get-gov.js:518]:
```suggestion
      // in JS (completely remove from the DOM) so we update the headers/labels based on total number of forms.
```
Flashback to pre 2012 internet

**Review by zandercymatics — COMMENTED** [review]
The code LGTM, minus one concern with the JS but its not blocking. Moving onto running the code now, and will come back with an approval if all is well

File: src/registrar/assets/sass/_theme/_forms.scss [hunk 5]
```diff
@@ -31,3 +31,10 @@
   padding-left: 0;
   border-left: none;
 }
+
+legend.float-left-tablet + button.float-right-tablet {
+  margin-top: .5rem;
+  @include at-media('tablet') {
```
Comment (therealslimhsiehdy) [inline src/registrar/assets/sass/_theme/_forms.scss:37]: **Question:**  `@include at-media('media--lg')` pattern instead?

**Review by therealslimhsiehdy — APPROVED** [review]
Approval with the caveat that I'd have a second person more well versed in JS to approve - from my limited knowledge and manual testing it looks great, but they may catch some edge cases or syntax I am not aware of.

Otherwise looks great in both desktop and phone view and love that when you delete the first Organization contact is moves the 2nds info to the 1st!! 

I do have a question which can be opened as a bug if it is one:
It seems also that if we require at least 1 organization contact, but for Organization contact 2 we still have required markers (*) on them. The field is required, but if they already have 1, should the second option still have required markers?

Screenshot 2024-01-11 at 11 13 28 AM

**Review by zandercymatics — APPROVED** [review]
LGTM, really nice work. Super smooth and streamlined. The only issue I could find is that ANDI doesn't seem to mark the delete button in the tab order, but I can tab onto it just fine and use it normally. I think its a non-issue.

I found a bug but its not related to this PR, but the preceding one. This probably doesn't need to be fixed here, but if you tab onto the radio field and hit enter to make a selection, the page will not let you proceed. 
image

**Comment by therealslimhsiehdy** [comment]
> I do have a question which can be opened as a bug if it is one: It seems also that if we require at least 1 organization contact, but for Organization contact 2 we still have required markers (*) on them. The field is required, but if they already have 1, should the second option still have required markers?
> 
> Screenshot 2024-01-11 at 11 13 28 AM

I'm relooking at the Figma and it looks like this is the design that is requested.

**Comment by rachidatecs** [comment]
> Approval with the caveat that I'd have a second person more well versed in JS to approve - from my limited knowledge and manual testing it looks great, but they may catch some edge cases or syntax I am not aware of.
> 
> Otherwise looks great in both desktop and phone view and love that when you delete the first Organization contact is moves the 2nds info to the 1st!!
> 
> I do have a question which can be opened as a bug if it is one: It seems also that if we require at least 1 organization contact, but for Organization contact 2 we still have required markers (*) on them. The field is required, but if they already have 1, should the second option still have required markers?
> 
> Screenshot 2024-01-11 at 11 13 28 AM

That's how we're handling DNSSEC. If you 'add' a form, you gotta fill it up. Or remove it.

**Comment by rachidatecs** [comment]
> LGTM, really nice work. Super smooth and streamlined. The only issue I could find is that ANDI doesn't seem to mark the delete button in the tab order like the other clickable elements, but I can tab onto it just fine and use it normally. I think its a non-issue.
> 
> I found a bug but its not related to this PR, but the preceding one. This probably doesn't need to be fixed here, but if you tab onto the radio field and hit enter to make a selection, the page will not let you proceed. image

The page is not supposed to let you proceed, you need to fill in the rationale for no other contacts. Hitting enter selects a radio option OK for me.

File: src/registrar/templates/application_other_contacts.html [hunk 6]
```diff
@@ -74,6 +88,9 @@ <h2>Organization contact {{ forloop.counter }} (optional)</h2>
           <legend>
             <h2>No other employees from your organization?</h2>
           </legend>
+          <p>You don't need to provide names of other employees now, but it may
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/templates/application_other_contacts.html:91]: Good point. @cisagov/gov-designers

File: src/registrar/assets/js/get-gov.js [hunk 7]
```diff
@@ -229,127 +229,245 @@ function handleValidationClick(e) {
...
+
+/**
+ * Prepare the namerservers, DS data and Other Contacts formsets' delete button
+ * for the last added form. We call this from the Add function
+ * 
+ */
+function prepareNewDeleteButton(btn, formLabel) {
+  let formIdentifier = "form"
+  let isNameserversForm = document.title.includes("DNS name servers |");
+  let isOtherContactsForm = document.title.includes("Other employees from your organization");
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/assets/js/get-gov.js:362]: Great point

File: src/registrar/assets/sass/_theme/_forms.scss (hunk 5 above)
Comment (rachidatecs) [reply src/registrar/assets/sass/_theme/_forms.scss:37]: I've never seen this token before. Do you have a link to docs?

In any case, we've used the mobile, tablet and desktop tokens for USWDS media query mixin so we either change all or stick with our convention.

Comment (therealslimhsiehdy) [reply src/registrar/assets/sass/_theme/_forms.scss:37]: Ah my apologies, I smooth brained -- I meant `@include at-media(mobile-lg)`. That's the more common one I've been using, but I do see both `tablet` and `mobile-lg` used on `getgov-home` and `manage.get.gov` so no fuss on what y'all already have!

**Comment by zandercymatics** [comment]
> The page is not supposed to let you proceed, you need to fill in the rationale for no other contacts. Hitting enter selects a radio option OK for me.

On my end I just get a "This field is required" error and that additional information won't appear (i.e. I can't enter any additional information, as I am stuck on the radio button selector - there is no textbox to begin with). Weird that its working for you though. I wonder if its a local config thing or something, what browser did you use - I would assume chrome?

**Comment by therealslimhsiehdy** [comment]
Hey big ask if it's possible to sneak into this PR (I chatted with Vicky about it), or we can also create a ticket if you don't feel comfortable:
 
 When one fills out the "Purpose of your domain" in the Domain Application - if you leave it blank and click "Save and continue" you will get red alerts and errors, meaning that this field seems to be required.

Is is possible to add in a line underneath "read about..." saying `Required fields are marked with an asterisk (*).` in italics like it for other parts of the form?

Screenshot 2024-01-12 at 9 41 32 AM

Screenshot 2024-01-12 at 9 19 59 AM


Let me know if you're okay with sneaking it in, or if you'd prefer us to open a ticket. No fuss on either choices!

**Comment by Katherine-Osos** [comment]
@rachidatecs @therealslimhsiehdy Forms that only have one field should **not** show the "required" messaging. However, if that solitary field is optional, we note that. @kristinacyin made this update to follow USWDS guidelines.

File: src/registrar/templates/application_other_contacts.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (gabydisarli) [reply src/registrar/templates/application_other_contacts.html:91]: following up. We're going off guidance from USWDS that forms with only one field don't need to be marked as required.

File: src/registrar/templates/application_other_contacts.html [hunk 8]
```diff
@@ -74,6 +88,9 @@ <h2>Organization contact {{ forloop.counter }} (optional)</h2>
           <legend>
             <h2>No other employees from your organization?</h2>
```
Comment (gabydisarli) [inline src/registrar/templates/application_other_contacts.html:89]: @rachidatecs can we remove the bottom margin here of 8px. There is already 16ishpx of margin above the body below this question and it feels too spaced out.

**Comment by rachidatecs** [comment]
> On my end I just get a "This field is required" error and that additional information won't appear (i.e. I can't enter any additional information, as I am stuck on the radio button selector - there is no textbox to begin with). Weird that its working for you though. I wonder if its a local config thing or something, what browser did you use - I would assume chrome?

I think I see what's going on. You are submitting the form without a radio button selection. In that case, the error you're getting makes sense: make a selection.

After you make a selection, the second part of the form appears. If you submit without filling out that second part, you get an error again, and that too makes sense.

If I misunderstood, let's huddle next week.

**Comment by rachidatecs** [comment]
We could customize the error message for the Yes/No initial form to say something like "You must make a selection."

@Katherine-Osos?

File: src/registrar/templates/application_other_contacts.html (hunk 8 above)
Comment (gabydisarli) [reply src/registrar/templates/application_other_contacts.html:89]: I can't seem to find it in this file, but can you change `<fieldset class="usa-fieldset margin-top-2">` to have a top margin of rem 2?

I don't know if I'm referring to it correctly, but looking at this `_utility-builder.scss` in the styles panel it is showing 1 rem.

Comment (gabydisarli) [reply src/registrar/templates/application_other_contacts.html:89]: In human words, the margin above "No other employees from your organization"

**Comment by gabydisarli** [comment]
> We could customize the error message for the Yes/No initial form to say something like "You must make a selection."
> 
> @Katherine-Osos?

Hey @rachidatecs I'll be doing this design review so I can help. What is the issue?

**Comment by gabydisarli** [comment]
> We could customize the error message for the Yes/No initial form to say something like "You must make a selection."
> 
> @Katherine-Osos?

I think I'm understanding now. You're suggesting to update the error if you don't select either option and then try to proceed?

**Comment by gabydisarli** [comment]
2 changes:

1. Please change the padding of the first usa-form-group to 0 rem instead of 0.5 rem. I want the titles of the org contact fields to be closer to the fields themselves. So removing that padding above the first field of each group helps.
2. Please change that error message when no radio is selected to: "This question is required."

I'm continuing the review but wanted to get these in to keep things moving

**Review by gabydisarli — APPROVED** [review]
LGTM!

**Pull request merged**